numpy/core/src/multiarray/datetime.c

```diff
@@ -473,169 +486,100 @@ convert_datetime_to_datetimestruct(PyArray_DatetimeMetaData *meta,
         case NPY_FR_h:
             perday = 24LL;
 
-            if (dt >= 0) {
-                set_datetimestruct_days(dt / perday, out);
-                dt  = dt % perday;
-            }
-            else {
-                set_datetimestruct_days((dt - (perday-1)) / perday, out);
-                dt = (perday-1) + (dt + 1) % perday;
-            }
+            set_datetimestruct_days(extract_unit(&dt, perday), out);
             out->hour = (int)dt;
             break;
 
         case NPY_FR_m:
             perday = 24LL * 60;
 
-            if (dt >= 0) {
-                set_datetimestruct_days(dt / perday, out);
-                dt  = dt % perday;
-            }
-            else {
-                set_datetimestruct_days((dt - (perday-1)) / perday, out);
-                dt = (perday-1) + (dt + 1) % perday;
-            }
-            out->hour = (int)(dt / 60);
-            out->min = (int)(dt % 60);
+            set_datetimestruct_days(extract_unit(&dt, perday), out);
+            out->hour = (int)extract_unit(&dt, 60);
+            out->min = (int)dt;
             break;
 
         case NPY_FR_s:
             perday = 24LL * 60 * 60;
 
-            if (dt >= 0) {
-                set_datetimestruct_days(dt / perday, out);
-                dt  = dt % perday;
-            }
-            else {
-                set_datetimestruct_days((dt - (perday-1)) / perday, out);
-                dt = (perday-1) + (dt + 1) % perday;
-            }
-            out->hour = (int)(dt / (60*60));
-            out->min = (int)((dt / 60) % 60);
-            out->sec = (int)(dt % 60);
+            set_datetimestruct_days(extract_unit(&dt, perday), out);
+            out->hour = (int)extract_unit(&dt, 60*60);
+            out->min  = (int)extract_unit(&dt, 60);
+            out->sec  = (int)dt;
```

Comment:
Looks even tidier as:
```
out->sec  = (int)extract_unit(&dt, 60);
out->min  = (int)extract_unit(&dt, 60);
out->hour = (int)extract_unit(&dt, 24);
set_datetimestruct_days(dt, out);
```
But that's less efficient, as the `dt < 0` branch can't be optimized out in the non-first calls to `extract_unit`

Reply:
I'm not too worried about efficiency just yet myself--we both know there's a lot more work to do still for correctness.

My only reservation with refactoring / abstracting is that we can't trust our current tests to catch subtle (and even breaking) changes just yet.